### PR TITLE
[training migration] Add LoggerConfig dataclass

### PR DESCRIPTION
**Pull request by maanug-nv**
# What does this PR do ?

1. Create dataclass in MLM
2. Get MLM arguments to use dataclass

:warning: For major changes (either in lines of code or in its impact), please make sure to first share discuss a design-doc with the team.  

## Contribution process

```mermaid
flowchart LR
    A[Pre-checks] --> B[PR Tests]
    subgraph Code Review/Approval
        C1[Expert Review] --> C2[Final Review]
    end
    B --> C1
    C2 --> D[Merge]
```

### Pre-checks

- [ ] I want this PR in a versioned release and have added the appropriate Milestone (e.g., `Core 0.8`)
- [ ] I have added relevant unit tests
- [ ] I have added relevant functional tests
- [ ] I have added proper typing to my code [Typing guidelines](https://docs.python.org/3/library/typing.html)
- [ ] I have added relevant documentation
- [ ] I have run the [autoformatter.sh](https://github.com/NVIDIA/Megatron-LM/blob/main/tools/autoformat.sh) on my PR

### Code review

The following process is enforced via the CODEOWNERS file for changes into `megatron/core`. For changes outside of `megatron/core`, it is up to the PR author whether or not to tag the Final Reviewer team.

<details>
<summary>For MRs into `main` branch</summary>

#### (Step 2): Collect the expert reviewers reviews

1. Attach the `Expert Review` label when your PR is ready for review.
2. GitHub auto-assigns expert reviewers based on your changes. They will get notified and pick up your PR soon.

:warning: Only proceed to the next step once all reviewers have approved, merge-conflict are resolved and the CI is passing.  
Final Review might get declined if these requirements are not fulfilled.

#### (Step 3): Final Review

1. Add `Final Review` label
2. GitHub auto-assigns final reviewers based on your changes. They will get notified and pick up your PR soon.

#### (Optional Step 4): Cherry-pick into release branch

If this PR also needs to be merged into `core_r*` release branches, after this PR has been merged, select `Cherry-pick` to open a new PR into the release branch.

</details>

<details>
<summary>For MRs into `dev` branch</summary>
The proposed review process for `dev` branch is under active discussion.

MRs are mergable after one approval by either `eharper@nvidia.com` or `zijiey@nvidia.com`.
</details>

### Merging your PR

Any member of [core-adlr](https://github.com/orgs/teams/NVIDIA/core-adlr) and [`core-nemo`](https://github.com/orgs/teams/NVIDIA/core-nemo) will be able to merge your PR.

### CODE REVIEW

File: megatron/training/argument_utils.py [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+import enum
+from dataclasses import Field, fields
+
+# TODO: support arg renames
+
+class TypeInferenceError(Exception):
+    """Custom exception type to be conditionally handled by ArgumentGroupFactory."""
+    pass
+
+class ArgumentGroupFactory:
```
Comment (Skylion007) [inline megatron/training/argument_utils.py:21]: This is basically reinventing hydra lol with slightly different syntax.

If you pass an arg parse object into OmegaConf.fromdotlist it should be equivalent

Comment (maanug-nv) [reply megatron/training/argument_utils.py:21]: `OmegaConf.from_dotlist()` requires a list of strings, so I'm not sure how you could pass an `argparse.Namespace` into it?

Intent here is to preserve the argparse interface and arguments for MegatronLM, since many users have scripts that are built on top of that (i.e., cannot switch to something like hydra). Using dataclasses in the training loop presents several benefits, but I'd like to avoid having to define the argument in 2 places, once in the dataclass definition and once in `arguments.py` with `add_argument()`.

File: megatron/training/arguments.py [hunk 2]
```diff
@@ -1930,83 +1930,11 @@ def _add_config_logger_args(parser):
...
-                       'Ignore wandb by default.')
-    group.add_argument('--wandb-exp-name', type=str, default='',
-                       help='The wandb experiment name.')
-    group.add_argument('--wandb-save-dir', type=str, default='',
-                       help='Path to save the wandb results locally.')
-    group.add_argument('--logging-level', type=int, default=None,
-                       help='Set default logging level')
+    from megatron.training.training_config import LoggerConfig
+
+    log_factory = ArgumentGroupFactory(LoggerConfig, exclude = ["log_throughput_to_tensorboard", "throughput_window_size", "memory_keys", "log_l2_norm_grad_to_tensorboard", "log_runtime_to_tensorboard", "runtime_time_unit", "filter_warnings", "modules_to_filter", "set_level_for_all_loggers", "save_config_filepath"])
```
Comment (maanug-nv) [inline megatron/training/arguments.py:2030]: Several of these excluded settings are for metrics currently logged in Megatron Bridge but not in MegatronLM. I think we will eventually like to upstream logging of those metrics to MegatronLM.
cc @dimapihtar

File: megatron/training/arguments.py [hunk 3]
```diff
@@ -2025,83 +2025,11 @@ def _add_config_logger_args(parser):
...
-                       '   1: report timing for operations that are executed '
-                       '      very limited times (basically once) during '
-                       '      each iteration (such as gradient all-reduce) '
-                       '   2: report timing for operations that migh be '
-                       '      executed numerous times during each iteration. '
-                       'Note that setting the level to 1 or 2 might '
-                       'cause increase in iteration time.')
-    group.add_argument('--log-energy', action='store_true',
-                       help='If set, log energy consumption (in Joules)')
-    group.add_argument('--no-barrier-with-level-1-timing', action='store_false',
```
Comment (Wohox) [inline megatron/training/arguments.py:2054]: Please correct me if I am wrong:
`--no-barrier-with-level-1-timing` defaults to True, which means `barrier_with_level_1_timing` defaults to False, which is not aligned in LoggerConfig.
* If user scripts doesn't contain `--no-barrier-with-level-1-timing`, then the default behavior changes.
* If user scripts contains `--no-barrier-with-level-1-timing`, then behavior is the same.

Comment (jaredcasper) [reply megatron/training/arguments.py:2054]: My understanding:

The action here is `store_false`, which means the presence of this argument sets `barrier_with_l1_timing` (the `dest`) to false. Without this flag (i.e. default), `barrier_with_l1_timing` is True. 

But it does seems like the LoggerConfig gets it wrong, unless it somehow knows the argument's action should be `store_false` (which happens in a few places in our arguments).

Comment (maanug-nv) [reply megatron/training/arguments.py:2054]:
```
from argparse import ArgumentParser


if __name__ == "__main__":
    parser = ArgumentParser()
    parser.add_argument('--no-barrier-with-level-1-timing', action='store_false', dest='barrier_with_L1_time')
    args = parser.parse_args()

    print(args)
```

```
maanug:~ ➜ python test-l1.py
Namespace(barrier_with_L1_time=True)
maanug:~ ➜ python test-l1.py --no-barrier-with-level-1-timing
Namespace(barrier_with_L1_time=False)
```

If user script does not contain the flag default is True. If user script contains the flag, changes to False.
More or less confirming what @jaredcasper said.

`ArgumentGroupFactory` sets `store_false` if default is True and `store_true` if default is False.
https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/training/argument_utils.py#L166-L172

Comment (maanug-nv) [reply megatron/training/arguments.py:2054]: Here are the kwargs for add_argument() for this one:
```
{'arg_names': ['--no-barrier-with-level-1-timing'], 'dest': 'barrier_with_L1_time', 'help': 'If not disabled, use barrier with level 1 time measurements. Note that this is up to the user to make sure calling barrier with their timers will not result in hangs. This can happen if for example the user adds a level 1 timer that is not called by all ranks.', 'default': True, 'action': 'store_false'}
```

i'm just printing [this](https://github.com/NVIDIA/Megatron-LM/blob/main/megatron/training/argument_utils.py#L202) and running `pretrain_gpt.py --help`
